src/drivers/i2c/i2c.h

```diff
@@ -1,27 +1,53 @@
 #ifndef DRIVERS_I2C_H_
 #define DRIVERS_I2C_H_
 
-struct i2c_device {
-	struct i2c_device *next;
+#include <stdint.h>
+
+#include <framework/mod/options.h>
+#include <config/embox/driver/i2c.h>
+
+#define I2C_BUS_MAX   \
+	OPTION_MODULE_GET(embox__driver__i2c, NUMBER, i2c_bus_max)
+
+#define MAX_I2C_BUS_NAME 8
+
+struct i2c_algorithm;
+
+struct i2c_msg {
+	uint16_t addr;           /* slave address */
+	uint16_t flags;
+	/* I2C_M_RD is guaranteed to be 0x0001! */
+#define I2C_M_RD    0x0001  /* read data, from slave to master */
+	uint16_t len;           /* msg length  */
+	uint8_t *buf;           /* pointer to msg data */
 };
 
 struct i2c_adapter {
-	void *priv;
+	void *i2c_algo_data;
+	const struct i2c_algorithm *i2c_algo; /* the algorithm to access the bus */
+};
+
+struct i2c_algorithm {
+	int (*i2c_master_xfer)(struct i2c_adapter *adap, struct i2c_msg *msgs,
+			int num);
 };
 
 struct i2c_bus {
 	int id;
-	int free_entry;
-	struct i2c_adapter *adapter;
-	struct i2c_device devices;
+	char name[MAX_I2C_BUS_NAME];
+
+	struct i2c_adapter *i2c_adapter;
 };
 
-int register_i2c_bus(struct i2c_adapter*);
-int unregister_i2c_bus(int bus_id);
-struct i2c_bus* get_i2c_bus(int);
-int enumerate_i2c_buses(int (*)(int, void*), void*);
+extern int i2c_bus_register(struct i2c_adapter *adap, int id,
+		const char *bus_name);
+
+extern int i2c_bus_unregister(int bus_id);
+
+extern struct i2c_bus *i2c_bus_get(int id);
+
+extern int i2c_bus_read(int id, uint16_t addr, const uint8_t *ch, size_t sz);
```

Comment:
Why use `const uint8_t *ch` instead of `uint8_t *ch`? I guess read function should be able modify buffer.